text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,660 @@
+  | Ambient                           |
+  +-+-----------+---------------------+         +------------+
+    | Create m1 | . . . . . . . . . . . . . . . | Receive m1 |
+    +-----------+-----------+                   +------------+
+                | Create m2 | . . . . . . . .
+                +-----------+---------+     .   +------------+
+                            | Publish |     . . | Receive m2 |
+                            +---------+         +------------+
+```
+
+#### Single message producers, batch push-based consumer with process spans
+
+```
+  PRODUCER                         CONSUMER
+
+  +------------+
+  | Publish m1 | . . . . . . . . . . . . . .
+  +------------+                           .
+                                   +---------------------------+
+                               . . | Deliver                   |
+  +------------+               .   +-+------------+------------+
+  | Publish m2 | . . . . . . . .     | Process m1 |
+  +------------+                     +------------+------------+
+                                                  | Process m2 |
+                                                  +------------+
+```
+
+#### Single message producers, batch pull-based consumer with process spans
+
+```
+  PRODUCER                    CONSUMER
+
+  +------------+
+  | Publish m1 |. . . .       +-------------------------------------+
+  +------------+      .       | Ambient                             |
+                      .       +-+---------+-------------------------+
+                      . . . . . | Receive |
+  +------------+          .     +---------+------------+
+  | Publish m2 |. . . . . .               | Process m1 |
+  +------------+                          +------------+------------+
+                                                       | Process m2 |
+                                                       +------------+
+```
+
+#### Single message producers, batch pull-based consumer with manual settlement
+
+```
+  PRODUCER                    CONSUMER
+
+  +------------+
+  | Publish m1 |. . . .       +----------------------------------------+
+  +------------+      .       | Ambient                                |
+                      .       +-+---------+----------------------------+
+                      . . . . . | Receive |
+  +------------+          .     +---------+    +-----------+
+  | Publish m2 |. . . . . .                    | Settle m1 |
+  +------------+                               +-----------+-----------+
+                                                           | Settle m2 |
+                                                           +-----------+
+```
+
+## Future possibilities
+
+### Transport context propagation
+
+One possibility to seamlessly integrate producer/consumer and intermediary
+instrumentation in a flexible and extensible way would be the introduction of a
+second transport context layer in addition to the creation context layer.
```

Comment:
With the presence of two contexts in a message, how would we envision usage of Propagators to inject and extract two different contexts from a single message?

Should messaging providers implement different getters/setters for each type of context?

Another approach is that when injecting a context, it is always the creation context if it doesn't exist (the first context injected context is always the creation context). If the creation context exists, then an injected context is implicitly the transport context. On extraction, assume it is always the transport context being extracted (where the creation context *is* the transport context if the transport context isn't explicitly present). This means the only problem is how to extract the creation context. If it is only auto-instrumented SDK's that will use this context in order to link a consumer span with a producer span, then perhaps this can be an implementation detail within the SDK that doesn't need to be exposed via a public interface?

I know some of these details may feel out of scope, but if the intent is to allow the current spec to allow for transport tracing to occur without a prescription for exactly how it's done, and if the intent is to introduce it in a later revision without breaking what is currently in the spec, we should be able to answer these questions now.

Reply:
Thanks for bringing this OTEP to a wider audience and for bringing up those questions. The following is my understanding.

We made the requirement, that the creation is attached to a message in a way so that it is not possible to be changed by intermediaries. This makes it like a permanent "stamp" on the message: regardless of what propagators intermediaries use and regardless of how they use it, the creation context should never change.

> Another approach is that when injecting a context, it is always the creation context if it doesn't exist

This is more of a gut feeling, but I think there have to be different kinds of propagators for creation and transport context. Currently existing propagators extract and inject transport context. This can be extended by having an additional propagator that extracts and injects creation context. Every component/intermediary that is messaging-agnostic can use standard propagators that work on the transport context (e. g. HTTP web servers). Components that are messaging-aware can additionally use propagators that work on the creation context.

> I know some of these details may feel out of scope, but if the intent is to allow the current spec to allow for transport tracing to occur without a prescription for exactly how it's done, and if the intent is to introduce it in a later revision without breaking what is currently in the spec, we should be able to answer these questions now.

Yes, anything we come up with should not block the introduction of intermediary instrumentation. One good way to eliminate concerns would be to come up with a rough prototype that shows how intermediary instrumentation could be done. I came up with a prototype some time ago, it might be worth extending that by looking at how propagators could be used.